### PR TITLE
Github Actions for building with shared libs && reorganized CI order

### PR DESCRIPTION
**Pull request by lwilkovich**
*Description of changes:*
Added actions for building artifacts with the shared_libs turned on.
Reordered the actions.
This also required some small changes to the build script.
Passing tests can be seen in the actions tab

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/build.sh [hunk 1]
```diff
@@ -111,6 +127,9 @@ case $compileMode in
       # Fix for the Cmake executing build of the sdk which errors out linking incorrectly to openssl
       cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_TOOLCHAIN_FILE=../cmake-toolchain/Toolchain-armhf.cmake -DBUILD_SDK=ON -DEXCLUDE_JOBS=ON -DEXCLUDE_DD=ON -DEXCLUDE_FP=ON -DDISABLE_MQTT=ON ../ || true
       cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_TOOLCHAIN_FILE=../cmake-toolchain/Toolchain-armhf.cmake -DBUILD_SDK=ON -DEXCLUDE_JOBS=ON -DEXCLUDE_DD=ON -DEXCLUDE_FP=ON -DDISABLE_MQTT=ON ../
+    elif [ "$sharedLibs" = true ]; then
+      cmake -DBUILD_SHARED_LIBS=ON -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_TOOLCHAIN_FILE=../cmake-toolchain/Toolchain-armhf.cmake -DBUILD_SDK=ON ../ || true
```
Comment (matthewpaul) [inline .github/build.sh:131]: Can you file a ticket for addressing the duplicate cmake calls in this file?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -59,9 +37,97 @@ jobs:
...
+      - name: Build ${{ env.PACKAGE_NAME }}
+        run: |
+          echo "${{ secrets.GITHUB_TOKEN }}" | docker login docker.pkg.github.com -u awslabs --password-stdin
+          export DOCKER_IMAGE=docker.pkg.github.com/awslabs/aws-iot-device-client/ubuntu-16-x64:latest
+          docker pull $DOCKER_IMAGE
+          docker run --mount type=bind,source=$(pwd),target=/root/${{ env.PACKAGE_NAME }} --env GITHUB_REF $DOCKER_IMAGE --compiler=clang-${{ matrix.version }}
+
   build-armhf32:
     runs-on: ubuntu-latest
-    if: github.ref == 'refs/heads/main'
```
Comment (matthewpaul) [inline .github/workflows/ci.yml:64]: Why are you removing this conditional check? Won't this increase the status check times for all branches? And is there an account level limit for number of actions workflows we can run?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -137,11 +201,33 @@ jobs:
...
+      - name: Archive ubuntu-16-x64 Build Artifact                                                                                                                                                        
+        uses: actions/upload-artifact@v2                                                                                                                                                                      
+        # Run for main branch only                                                                                                                                                                            
+        if: github.ref == 'refs/heads/main'                                                                                                                                                                   
+        with:                                                                                                                                                                                                 
+          name: "SL.ubuntu.x64.${{ env.PACKAGE_NAME }}"
```
Comment (matthewpaul) [inline .github/workflows/ci.yml:212]: I think this naming convention is confusing - the static builds get DC for Device Client but the shared library builds get SL for shared library. I think all of them should get DC to differentiate between ST builds and then somewhere in the artifact name you can just append "shared", with static builds just not having the extension, or having "static"

File: .github/workflows/ci.yml (hunk 2 above)
Comment (lwilkovich) [reply .github/workflows/ci.yml:64]: There is an account level limit for the number of workflows at a given time I believe.  And the number of jobs in a workflow is limited by time, where we're no where near the cap.  I also felt that we should be building these each time for more transparency, it doesn't really add much time to wait.

**Review by matthewpaul — APPROVED** [review]

**Review by HarshGandhi-AWS — APPROVED** [review]

**Pull request merged**